Suppress false positive warning about uninitialized entropy buffers

Sjors opened the pull request:
The memory sanitizer cannot see through `syscall` (see https://github.com/google/sanitizers/issues/852). This caused a false
positive `use-of-uninitialized-value` error for `SYS_getrandom`.

Given recent problematic real issues with uninitialized variables, I think it's worth getting rid of the false positives so developers, and maybe Travis, can run the sanitizer.

```
==18287==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0x561f1c40bfc8 in ReadBE64(unsigned char const*) /home/dev/bitcoin/src/./crypto/common.h:67:12
    #1 0x561f1c3fe1d3 in (anonymous namespace)::sha512::Transform(unsigned long*, unsigned char const*) /home/dev/bitcoin/src/crypto/sha512.cpp:59:63
    #2 0x561f1c3fcb77 in CSHA512::Write(unsigned char const*, unsigned long) /home/dev/bitcoin/src/crypto/sha512.cpp:168:9
    #3 0x561f1c2d102d in CSHA512& (anonymous namespace)::operator<<<timespec>(CSHA512&, timespec const&) /home/dev/bitcoin/src/randomenv.cpp:119:12
    #4 0x561f1c2d07e8 in RandAddDynamicEnv(CSHA512&) /home/dev/bitcoin/src/randomenv.cpp:249:12
    #5 0x561f1c2cbfb9 in SeedStartup(CSHA512&, (anonymous namespace)::RNGState&) /home/dev/bitcoin/src/random.cpp:513:5
    #6 0x561f1c2c933e in ProcRand(unsigned char*, int, RNGLevel) /home/dev/bitcoin/src/random.cpp:553:9
    #7 0x561f1c2c8f98 in GetRandBytes(unsigned char*, int) /home/dev/bitcoin/src/random.cpp:558:59
    #8 0x561f1baa521e in (anonymous namespace)::CSignatureCache::CSignatureCache() /home/dev/bitcoin/src/script/sigcache.cpp:34:9
    #9 0x561f1b3bcdf9 in __cxx_global_var_init.3 /home/dev/bitcoin/src/script/sigcache.cpp:67:24
    #10 0x561f1b3bce88 in _GLOBAL__sub_I_sigcache.cpp /home/dev/bitcoin/src/script/sigcache.cpp
    #11 0x561f1c60107c in __libc_csu_init (/home/dev/bitcoin/src/bitcoind+0x129507c)
    #12 0x7fc924a8fb27 in __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:266
    #13 0x561f1b3c3659 in _start (/home/dev/bitcoin/src/bitcoind+0x57659)

  Uninitialized value was stored to memory at
    #0 0x561f1b3c9316 in __msan_memcpy (/home/dev/bitcoin/src/bitcoind+0x5d316)
    #1 0x561f1c3fcc8c in CSHA512::Write(unsigned char const*, unsigned long) /home/dev/bitcoin/src/crypto/sha512.cpp:179:9
    #2 0x561f1c2cbc50 in SeedSlow(CSHA512&) /home/dev/bitcoin/src/random.cpp:467:12
    #3 0x561f1c2cbf65 in SeedStartup(CSHA512&, (anonymous namespace)::RNGState&) /home/dev/bitcoin/src/random.cpp:509:5
    #4 0x561f1c2c933e in ProcRand(unsigned char*, int, RNGLevel) /home/dev/bitcoin/src/random.cpp:553:9
    #5 0x561f1c2c8f98 in GetRandBytes(unsigned char*, int) /home/dev/bitcoin/src/random.cpp:558:59
    #6 0x561f1baa521e in (anonymous namespace)::CSignatureCache::CSignatureCache() /home/dev/bitcoin/src/script/sigcache.cpp:34:9
    #7 0x561f1b3bcdf9 in __cxx_global_var_init.3 /home/dev/bitcoin/src/script/sigcache.cpp:67:24
    #8 0x561f1b3bce88 in _GLOBAL__sub_I_sigcache.cpp /home/dev/bitcoin/src/script/sigcache.cpp
    #9 0x561f1c60107c in __libc_csu_init (/home/dev/bitcoin/src/bitcoind+0x129507c)

  Uninitialized value was created by an allocation of 'buffer' in the stack frame of function '_ZL8SeedSlowR7CSHA512'
    #0 0x561f1c2cbb80 in SeedSlow(CSHA512&) /home/dev/bitcoin/src/random.cpp:459

SUMMARY: MemorySanitizer: use-of-uninitialized-value /home/dev/bitcoin/src/./crypto/common.h:67:12 in ReadBE64(unsigned char const*)
Exiting
```

See #17620 for how to use the memory sanitizer.

laanwj commented:
I think in general, initializing everything *unless* it's deliberately not done for performance reasons, is a good thing.

Fixing random buffers because of valgrind false positives has some [very bad](https://research.swtch.com/openssl) precedent though.

practicalswift commented:
@laanwj To be fair this is not a Valgrind false positive? :)  Valgrind false positives are _very_ rare in my experience.

laanwj commented:
> To be fair this is not a Valgrind false positive? :) Valgrind false positives are very rare in my experience.

It's a false positive because the buffer is actually filled in by the syscall but valgrind doesn't notice. 

> The memory sanitizer cannot see through syscall (see google/sanitizers#852). This caused a false positive use-of-uninitialized-value error for SYS_getrandom.

practicalswift commented:
@laanwj Yes, but the false positive is in MemorySanitizer -- not Valgrind :)

elichai commented:
Concept ACK.
From a quick glance it seems like the only change is initializing the value, and because unlike the openssl/debian fiasco, we *don't* use unintialized memory for generating randomness then there's no logical reason why this can make anything that currently works break (unless it was already broken).

Still requires close up review but I think it's good to keep that in mind

Sjors commented:
Agree that it needs careful review. I can replace the `{0}` with "// intentionally not initialized" where needed.

Also note that the code this touches is relatively new (added by @sipa in #15250 and #17270), and deserves extra scrutiny anyway now that we've dropped the OpenSSL training wheels.

The actual initialization before this commit took place in `GetOSRand`, which is different depending on the operating system details. That's probably another good reason to make sure stuff is initialized before it goes in there.

practicalswift commented:
Concept ACK

> I think in general, initializing everything _unless_ it's deliberately not done for performance reasons, is a good thing. One company I've worked for even had a strict "initialize everything" policy, no exceptions.

Agree. I think it is worth doing.

Some background reading that might be of interest regarding the pros/cons of default initialization in general:
* C++ Core Guidelines (Sutter & Stroustrup), "Always initialize an object": https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#es20-always-initialize-an-object
* Clang automatic variable initialization: https://reviews.llvm.org/D54604 + https://twitter.com/jfbastien/status/1063203517102415872
* Apple has started using automatic variable initialization in iOS: https://twitter.com/josephbialek/status/1168930302036762625
* Microsoft has started shipping the Windows kernel with MSVC `InitAll` enabled: https://twitter.com/JosephBialek/status/1062774315098112001
* Talk "Making C Less Dangerous in the Linux kernel": https://www.youtube.com/watch?v=FY9SbqTO5GQ&feature=youtu.be&t=742
* PDF: "Fighting Uninitialized Memory in the Kernel": https://linuxplumbersconf.org/event/4/contributions/426/attachments/338/564/KMSAN_for_LPC_2019

Some quotes:
* Linus Torvalds: "So I'd like the zeroing of local variables to be a native compiler option, so that we can (_eventually_ - these things take a long time) just start saying "ok, we simply consider stack variables to be always initialized".
* Linus Torvalds: "Again - I don't think we want a world where everything is force-initialized. There _are_ going to be situations where that just hurts too much. But if we get to a place where we are zero-initialized by default, and have to explicitly mark the unsafe things (and we'll have comments not just about how they get initialized, but also about why that particular thing is so performance-critical), that would be a good place to be."

Note that there is a `clang-tidy` fixup that can rewrite this for us if we want (applies only to integers, booleans, floats, doubles and pointers unfortunately): https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-init-variables.html

practicalswift commented:
@Sjors Which subset of these changes are required to make MemorySanitizer happy? Not all these are technically needed, right? :)

Sjors commented on src/random.cpp:
```diff
@@ -451,7 +451,7 @@ static void SeedFast(CSHA512& hasher) noexcept
 static void SeedSlow(CSHA512& hasher) noexcept
 {
-    unsigned char buffer[32];
+    unsigned char buffer[32] = {0};
```
@practicalswift once I fixed this one, it ran into unrelated   `use-of-uninitialized-value` problems. That could mean this is the only problem, or the other ones just didn't get a chance to get called before the sanitizer bailed out.

jonatack commented:
I feel unqualified to ACK here, but read the changes here and the first 2 links provided by @practicalswift, built ccfed8a22536bc5a307cc463a64fc05c7aa18281 with gcc version 8.3.0 (Debian 8.3.0-6) on Debian 4.19.37-5+deb10u2 (2019-08-08), ran tests, and am running bitcoind without issues.

jonatack reviewed (commented):
Concept ACK. I feel unqualified to ACK here, but I read the changes and the first 2 links provided by @practicalswift, built ccfed8a without sanitizer (--enable-debug only) with gcc version 8.3.0 (Debian 8.3.0-6) on Debian 4.19.37-5+deb10u2 (2019-08-08), ran tests, and am running bitcoind without issues and seeing the `Feeding n bytes of dynamic environment data into RNG` logs with typical values for _n_ (~16700) on this installation.

gmaxwell commented:
Initializing this would conceal from valgrind if changes in the code cased the random variable to not get filled which would be a moderately serious problem: Zero is not an acceptable random value. :)

Since msan is actually incorrect due to a known and documented limitation in msan, might it not be better to use a suppression?  I'm not a fan of suppressing actual errors-- but this isn't an error, it's a tool limit, and suppressing-by-initializing will reduce the effectiveness other tools without the same limit.

elichai commented:
> Initializing this would conceal from valgrind if changes in the code cased the random variable to not get filled which would be a moderately serious problem: Zero is not an acceptable random value. :)
> 
> Since msan is actually incorrect due to a known and documented limitation in msan, might it not be better to use a suppression? I'm not a fan of suppressing actual errors-- but this isn't an error, it's a tool limit, and suppressing-by-initializing will reduce the effectiveness other tools without the same limit.

That's a good point. does valgrind support syscalls?

Sjors commented:
Maybe I can add `assert` statements directly after any buffers are filled  to check that they're not still 0?

practicalswift commented:
Clarification regarding my Concept ACK: a suppression would be totally fine too.

As long as we make it easier for everyone to use MemorySanitizer I'm happy: we need more MSAN testing :)

Sjors commented:
I added asserts to ensure these buffers are actually populated on first use. I also added @gmaxwell's comment about valgrind. I'm not very familiar with how valgrind works. If it substitutes calls to `hasher.Write()`, `hasher.Finalize()`, `GetOSRand()`, etc with no-op, then this won't work.

I could pick a magic value instead, initialize with that and use == in the assert.

laanwj commented:
> That's a good point. does valgrind support syscalls?

Yes, valgrind doesn't have this issue.

> Maybe I can add assert statements directly after any buffers are filled to check that they're not still 0? (though I suppose that could be forgotten in new places)

But what if the input is legitly 0? These are not random values being fed in, but just some OS data, which may or may not consist of zeroes.

So NACK on the asserts.

laanwj commented on src/random.cpp:
```diff
@@ -389,6 +391,8 @@ class RNGState {
             // Finalize the hasher
             hasher.Finalize(buf);
+            // Ensure Finalize() populated buf
+            assert(buf);
```
FWIW: asserting the *address* of the buffer, on the stack, like you do here does nothing.

Sjors replied:
Oops. Looping through it doesn't sound appealing either.

Sjors replied:
Yikes.

Sjors replied:
For future reference, something like this works (?), though it's hideous, probably needs a macro, which adds even more complexity:
```
unsigned char zeros[sizeof(buf)] = {0};
assert(!memcmp(fbuf, zeros, sizeof(buf)));
```

Sjors replied:
Funny who you run into while googling that: https://rusty.ozlabs.org/?p=560 :-)

laanwj commented:
I've thought about it a bit and I'm principally opposed to adding asserts here. Now you're changing *run-time* behavior of the application to work around a problem in an analysis tool.

I'm sorry to say this but : please fix the tool instead

practicalswift commented:
Echoing the NACK on the asserts :)

practicalswift commented:
@Sjors Have you considered taking the suppressions route instead? Perhaps that is less intrusive.

Sjors commented:
Reverted to 605991, which just initializes to zero. I'm not sure if intentially leaving them uninitialized and having the warnings suppressed if the right approach either.

practicalswift commented:
> I'm not sure if intentially leaving them uninitialized and having the warnings suppressed if the right approach either.

Why? :)

Sjors commented:
It seems like we're choosing between two tools here:
1. memory sanity check: `= {0}`
2. valgrind: `// intentionally undefined to allow valgrind to catch issues`

There's gotta be a solution that involves defined behavior. That would make the code easier to reason about, and prevents folks on different operating systems and compiler optimization flags not noticing bugs (as happened with  #17568 and #17449).

laanwj commented:
In that case, let's at least not penalize the tool that actually gets things right (valgrind).

Sjors commented:
Wouldn't initializing to a magic value (e.g `sha256("bitcoin")`) solve both valgrind and sanity check issue?

practicalswift commented:
@Sjors Why not doing a suppression? That seems to be the easiest way to solve the problem this PR is meant to solve :)

There are two ways to do MemorySanitizer suppresions: you can do the suppression in the code using the annotation `__attribute__((no_sanitize("memory")))` for the functions you want MSAN-suppressed, or you can use a sanitizer special case blacklist which is perhaps slightly more involved. See https://clang.llvm.org/docs/MemorySanitizer.html#usage for details :)

MarcoFalke commented:
NACK. We don't generally change code to please buggy tool or compiler warnings, especially if the "fix" hides actual bugs in our code base.

Sjors commented:
@MarcoFalke it's about more than making tools happy. The tool pointed us to these uninitialized variables, which have caused pretty serious bugs in the recent past and are bad practice. I don't think we should just ignore this.

practicalswift commented:
@Sjors You haven't addressed the question about suppressions. Why not suppress this case? :)

laanwj commented:
I agree with @MarcoFalke and @practicalswift  here.

As I said before, if we do anything here it should be a tool-specific suppression, until the tool is fixed.

Sjors commented:
I'm not against suppressing the warning, but I don't think that's enough by itself. It doesn't make the undefined behavior go away. Perhaps one day the address sanitizer can reason about undefined behavior even through syscall. That's great, but still not a good reason to keep the undefined behavior.

laanwj commented:
But it is not undefined behavior, *this specific tool* is unable to reason through this.

Sjors committed:
build: skip memory sanitizer for SeedSlow

Memory sanitizer cannot see through syscall. This caused a false
positive use-of-uninitialized-value error for SYS_getrandom.

After discussion in #17627 it was decided not to initialize
these variables to 0, but instead to suppress the warning.

Sjors commented:
I added the suppression, as narrowly as possible. I also added comments to the uninitialized variables, pointing back to this discussion so it doesn't get repeated.

MarcoFalke commented:
> @MarcoFalke it's about more than making tools happy. The tool pointed us to these uninitialized variables, which have caused pretty serious bugs in the recent past and are bad practice. I don't think we should just ignore this.

Again, this is a bug in the tool. These variables are not uninitialized when read. Initializing them with the wrong value will cause bugs in the future because tools that properly detect uninitialized reads will no longer yell at you.

This point is repeated every couple of months. Might be a good candidate to add to the dev notes.

MarcoFalke commented on src/random.cpp:
```diff
@@ -467,6 +467,13 @@ static void SeedEvents(CSHA512& hasher)
+#  if __has_feature(memory_sanitizer)
+__attribute__((no_sanitize("memory")))
+#  endif
+#endif
```
This is a clang bug, right? Should link to the commit that fixed it. And since you are compiling clang from source you might as well use that commit and then remove this code blob here.

Sjors replied:
The memory sanitizer appears to be a Google project, and they don't except `syscall` support to be fixed anytime soon. Only a more narrow case of `getrandom` was fixed, which is of no use to us: https://github.com/google/sanitizers/issues/852#issuecomment-327305535

MarcoFalke replied:
Ok, fine. I'd still prefer a suppression in a suppressions file if possible. Modifying the code like this makes it harder to read and potentially impossible to compile on some environments.

Sjors replied:
@practicalswift any idea how to achieve such a suppressions file?
The outer `#if defined(__has_feature)` is there to make sure other environments don't trip over it. I could put it in a macro, or we can do that if we ever need it again.

practicalswift replied:
@Sjors Sure! :) You need to use the sanitizer special case list functionality by passing `-fsanitize-blacklist=test/sanitizer_suppressions/msan` as part of `CFLAGS`/`CXXFLAGS`. You need to create the file `test/sanitizer_suppressions/msan` following the file format specified here: https://clang.llvm.org/docs/SanitizerSpecialCaseList.html#format

Sjors replied:
I prefer to punt on this until we need it in more places...

laanwj replied:
> I prefer to punt on this until we need it in more places...

So this is the only place where MSAN special treatment is needed? That's pretty good, I had assumed it would be the first of many.

Sjors replied:
I don't know yet. The binary exits once it finds a single problem. The next problem in line is in a different file, but who know what happens after that.

practicalswift commented:
@Sjors Thanks for using a suppression instead. Is this the only suppression needed to make MSAN happy in your setup?

Sjors commented:
@practicalswift no, it fails on other things, but those appear unrelated. I'll update #17620 later.

gmaxwell commented:
It's not clear to me that everyone understands what's going on here. And maybe I caused some confusion by say that zero is not an acceptable random value. No constant is an acceptable random value.

The variable is not undefined. It gets defined by the SYS_getrandom.  Memcheck has a flaw/shortcoming where it doesn't know or assume that sys_getrandom initializes it so it falsely reports it as uninitialized.

If it were actually undefined then that would mean that the getrandom call wasn't doing anything which would be a really serious problem because it would mean that the software getting random numbers it was expecting. Ignoring the other belt-and-suspenders randomness inputs, a failure to read random data into that variable would be a total break resulting in predictable keys.

With the code as it currently is, should some bug be added so that the variable is not initialized at least valgrind will start yelling.  If the code is changed to initialize the value with anything except a truly random value (such a zero or some random hash) then if a flaw that skips the randomness call is introduced the software will be just as vulnerable but in that case the flaw will be undetectable by valgrind. Similarly, compilers can often warn when they're certain a value will be undefined, and pre-emptive dummy initialization will break those warnings.

An assertion would not be appropriate because any value is acceptable random value, so long as it's actually random and not some predictable constant.

Another potential tool in the debate between "pre-initilize variables and avoid the risk of undefined values" and "don't pre-init and preserve valgrind/memcheck sensitivity to coding errors that cause the required "real initialization" is that valgrind.h has special macros that can be used to mark memory as undefined in valgrind's tracking although they're defined for language purposes. (VALGRIND_MAKE_MEM_UNDEFINED(x,y), see secp256k1/tests.c). I'm not sure if anything similar exists for memcheck.

In some rare cases pre-initilization is a performance concern. But in most-- where the variable isn't in some inner loop and doesn't involve a bunch of malloc-- it's not. It is, however, very often a testing sensitivity concern. Of course, in some cases the code can be re-factored to do the proper 'final' initialization at declaration time then that's best.

If there were a big swing in the project towards performing dummy initialization I think it would be really useful if it were always done via a macro that would allow disabling it for testing (or valgrind annotating it).

practicalswift commented:
> Memcheck has a flaw/shortcoming where it doesn't know or assume that sys_getrandom initializes it so it falsely reports it as uninitialized.

You mean MemorySanitizer (MSAN) and not Memcheck (Valgrind), right? Not Memcheck (Valgrind) :)

> In some rare cases pre-initilization is a performance concern. But in most-- where the variable isn't in some inner loop and doesn't involve a bunch of malloc-- it's not. It is, however, very often a testing sensitivity concern.

From a security perspective doesn't it boil down to a trade-off between maximising testing sensitivity (achieved by not pre-initializing) vs minimising the risk of leaking non-dummy memory content (achieved by pre-initializing)?

Is that a fair summary? :)

---

Background context for others:

Depending on code generation, a programmer who runs into undefined behavior by using an uninialized automatic variable may observe any previous value (including program secrets), or any value which the compiler saw fit to materialize on the stack or in a register (this could be to synthesize an immediate, to refer to code or data locations, to generate cookies, etc).

gmaxwell commented:
> minimising the risk of leaking non-dummy memory content (achieved by pre-initializing)

There is no risk of leaking anything here, and of the randomness is hashed before being used.

The difference between initializing and not is that if a bug is introduced: With initializing a bug that hobbles the random input will not be detectable (except via review),  without initializing the bug would be detectable by memcheck but there would also be undefined behaviour (however, on currently supported compilers/systems the undefined behaviour would be harmless esp compared to the no-randomness bug) until someone noticed the error report in memcheck and fixed it.

jonatack commented:
FWIW I attempted to summarise this discussion in the review club notes as a resource and will probably update it with the points made here. Don't hesitate to suggest corrections or improvements.  

https://bitcoincore.reviews/17639.html#what-can-we-do-to-mitigate-uninitialized-variables

practicalswift commented:
> > From a security perspective doesn't it boil down to a trade-off between maximising testing sensitivity (achieved by not pre-initializing) vs minimising the risk of leaking non-dummy memory content (achieved by pre-initializing)?
> 
> There is no risk of leaking anything here, and of the randomness is hashed before being used.

I was talking about the pros and cons of pre-initializing generally: in _this specific case_ there is obviously no risk of leaking sensitive data :)

Sjors commented:
Closing in favor of the more precise fix in #18288. We should still document a policy around initialisation, perhaps requiring _either_ initialisation or a comment that's intentionally not initialised.
